src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -234,15 +228,133 @@ static void CheckForExperimentStatus(ITextView textView, in CompletionOptions op
+                var (context, list) = await GetCompletionContextWorkerAsync(document, trigger, triggerLocation,
+                    options with { ExpandedCompletionBehavior = ExpandedCompletionMode.AllItems }, cancellationToken).ConfigureAwait(false);
+
+                AddPropertiesToSession(session, list, triggerLocation);
+                AsyncCompletionLogger.LogImportCompletionGetContext(isBlocking: true, delayed: false);
+                return context;
+            }
+            else
             {
-                return AsyncCompletionData.CompletionContext.Empty;
+                // OK, expand item is enabled but we shouldn't block completion on its results.
+                // Kick off expand item calculation first in background.
+                Stopwatch stopwatch = new();
+                var expandedItemsTask = Task.Run(async () =>
+                {
+                    var result = await GetCompletionContextWorkerAsync(document, trigger, triggerLocation,
+                      options with { ExpandedCompletionBehavior = ExpandedCompletionMode.ExpandedItemsOnly }, cancellationToken).ConfigureAwait(false);
+
+                    // Record how long it takes for the background task to complete *after* core providers returned.
+                    // If telemetry shows that a short wait is all it takes for ExpandedItemsTask to complete in
+                    // majority of the sessions, then we might consider doing that instead of return immediately.
+                    // There could be a race around the usage of this stopwatch, I ignored it since we just need a rough idea:
+                    // we always log the time even if the stopwatch's not started regardless of whether expand items are included intially
+                    // (that number can be obtained via another property.)
+                    AsyncCompletionLogger.LogAdditionalTicksToCompleteDelayedImportCompletionDataPoint((int)stopwatch.ElapsedMilliseconds);
+
+                    return result;
+                }, cancellationToken);
+
+                // Now trigger and wait for core providers to return;
+                var (nonExpandedContext, nonExpandedCompletionList) = await GetCompletionContextWorkerAsync(document, trigger, triggerLocation,
+                        options with { ExpandedCompletionBehavior = ExpandedCompletionMode.NonExpandedItemsOnly }, cancellationToken).ConfigureAwait(false);
+                AddPropertiesToSession(session, nonExpandedCompletionList, triggerLocation);
+
+                if (expandedItemsTask.IsCompleted)
+                {
+                    // the task of expanded item is completed, get the result and combine it with result of non-expanded items.
+                    var (expandedContext, expandedCompletionList) = await expandedItemsTask.ConfigureAwait(false);
+                    AddPropertiesToSession(session, expandedCompletionList, triggerLocation);
```

Comment:
Why do we call `AddPropertiesToSession` here when we already called it a few lines above?